File: LeetCode/LongestSubstringWithoutRepeatingCharacters.java
import java.lang.reflect.Array;
import java.util.Arrays;

/**
 * Created by yi on 4/15/15.
 */
public class LongestSubstringWithoutRepeatingCharacters {
    public int lengthOfLongestSubstring(String s) {
        int[] charmap = new int[256];
        Arrays.fill(charmap, -1);
        int i = 0;
        int maxl = 0;
        for(int j = 0; j < s.length(); j++) {
            if(charmap[s.charAt(j)] >= i)
                i = charmap[s.charAt(j)] + 1;
            charmap[s.charAt(j)] = j;
            maxl = Math.max(maxl, j - i + 1);
        }
        return maxl;
    }
}

File: LeetCode/Ruby/longest_substring_without_repeating_characters.rb
class LongestSubstringWithoutRepeatingCharacters
  def length_of_longest_substring(s)
    last = 0
    lookup = Hash(nil)
    (0...s.size).each do |i|
      if lookup.has_key?(s[i]) && lookup[s[i]] + 1 > last
        @max = [i - last, @max ? @max : 0].max
        last = lookup[s[i]] + 1
      end
      lookup[s[i]] = i
    end
    [@max, s.size - last].max

  end
end

puts LongestSubstringWithoutRepeatingCharacters.new
         .length_of_longest_substring("abba")